kernel/src/vm/vmar/mod.rs

```diff
@@ -422,41 +530,114 @@ impl Vmar_ {
         Ok(())
     }
 
-    // Split and unmap the found mapping if resize smaller.
-    // Enlarge the last mapping if resize larger.
-    fn resize_mapping(&self, map_addr: Vaddr, old_size: usize, new_size: usize) -> Result<()> {
-        debug_assert!(map_addr % PAGE_SIZE == 0);
-        debug_assert!(old_size % PAGE_SIZE == 0);
-        debug_assert!(new_size % PAGE_SIZE == 0);
-
-        if new_size == 0 {
-            return_errno_with_message!(Errno::EINVAL, "can not resize a mapping to 0 size");
+    /// Splits and unmaps the found mapping if the new size is smaller.
+    /// Enlarges the last mapping if the new size is larger.
+    fn resize_mapping(
+        &self,
+        map_addr: Vaddr,
+        old_size: usize,
+        new_size: usize,
+        check_single_mapping: bool,
+    ) -> Result<()> {
+        let mut rss_delta = RssDelta::new(self);
+        let mut inner = self.inner.write();
+        if check_single_mapping {
+            inner.check_lies_in_single_mapping(&(map_addr..map_addr + old_size))?;
         }
+        inner.resize_mapping(&self.vm_space, map_addr, old_size, new_size, &mut rss_delta)?;
+        Ok(())
+    }
 
-        if new_size == old_size {
-            return Ok(());
-        }
+    fn remap(
+        &self,
+        old_addr: Vaddr,
+        old_size: usize,
+        new_addr: Option<Vaddr>,
+        new_size: usize,
+    ) -> Result<Vaddr> {
+        let mut old_range = old_addr..old_addr.checked_add(old_size).ok_or(Errno::EINVAL)?;
+        let mut old_size = old_size;
+        let mut inner = self.inner.write();
+        let mut rss_delta = RssDelta::new(self);
 
-        let old_map_end = map_addr + old_size;
-        let new_map_end = map_addr + new_size;
+        let old_mapping_addr = inner.check_lies_in_single_mapping(&old_range)?;
```

Comment:
Following the above comment, this should live in the beginning of the method, so we can later use:
```rust
let old_range = old_addr..old_addr + old_size
```
without the overflowing check.